examples/threathunter/threat_intelligence/stix_taxii.py

```diff
@@ -0,0 +1,283 @@
+        if not self.client:
+            logging.error('Unable to create taxii client.  Exiting...')
+            return reports
+
+        available_collections = self.query_collections()
+        if not available_collections:
+            logging.warning('Unable to find any collections.  Exiting...')
+            return reports
+
+        reports = self.import_collections(available_collections)
+        if not reports:
+            logging.warning('Unable to import collections.  Exiting...')
+            return reports
+
+        return reports
+
+
+
+
+class StixTaxii():
+    def __init__(self, site_confs):
+        self.config = site_confs
+        self.client = None
+
+    def result(self, **kwargs):
+        """
+        Returns a new AnalysisResult with the given fields populated, updating
+        the database in the background.
+
+        This should be used within the :meth:`analyze` method to create
+        analysis results.
+
+        :rtype: :class:`AnalysisResult`
+
+        Example::
+
+        >>> self.result(analysis_name="foo", score=10)
+        """
+
+        result = AnalysisResult(**kwargs).normalize()
+        return result
+
+
+    def configure_sites(self):
+        self.sites = {}
+        try:
+            for site_name, site_conf in self.config['sites'].items():
+                self.sites[site_name] = TaxiiSiteConnector(site_conf)
+                logging.info(f"loaded site {site_name}")
+        except handled_exceptions as e:
+
+            logging.error(f"Error in parsing config file: {e}")
+
+    def format_report(self, reports):
+        for report in reports:
+            try:
+                analysis_name = report['id']
+                title = report['title']
+                description = report['description']
+                scan_time = datetime.fromtimestamp(report['timestamp'])
+                score = report['score']
+                link = report['link']
+                ioc_dict = report['iocs_v2']
+                result = self.result(
+                                     analysis_name=analysis_name,
+                                     scan_time=scan_time,
+                                     score=score,
+                                     title=title,
+                                     description=description)
+                for ioc_key, ioc_val in ioc_dict.items():
+                    result.attach_ioc_v2(values=ioc_val, field=ioc_key, link=link)
+            except handled_exceptions as e:
+                logging.warning(f"Problem in report formatting: {e}")
+                result = self.result(
+                    analysis_name="exception_format_report", error=True)
+            yield result
+
+    def analyze(self):
+        self.configure_sites()
+        ti = ThreatIntel()
+        for site_name, site_conn in self.sites.items():
+            logging.info(f"Talking to {site_name} server")
+            reports = site_conn.generate_reports()
+            if not reports:
```

Comment:
I think this function needs to be broken up since in one case you are returning a result and in the other you are pushing to Carbon Black Cloud. The confusion can be seen on line 283 where you pass in a for loop to push each individual report.

Reply:
Rewrote to just get and send, and also added more logical error reporting. Removed the need for a for loop to go through the generator.